luarules/gadgets/unit_projectile_leash.lua

```diff
@@ -0,0 +1,163 @@
+	local dx1 = originX - projectileX
+	local dz1 = originZ - projectileZ
+	local distanceToOrigin = mathSqrt(dx1 * dx1 + dz1 * dz1)
+
+	if distanceToOrigin > weaponRange then
+		if leashRange then
+			local distanceToOwner
+			local ownerX, ownerY, ownerZ = spGetUnitPosition(proOwnerID)
+			if ownerX then
+				local dx2 = ownerX - projectileX
+				local dz2 = ownerZ - projectileZ
+				distanceToOwner = mathSqrt(dx2 * dx2 + dz2 * dz2)
+				if distanceToOwner > leashRange then
+					return true
+				end
+			end
+			return false
+		end
+		return true
+	end
+	return false
+end
+
+local function projectileIsCloseToEdge(rangeThreshold, originX, originZ, projectileX, projectileZ)
+	local dx1 = originX - projectileX
+	local dz1 = originZ - projectileZ
+	local distanceToOrigin = mathSqrt(dx1 * dx1 + dz1 * dz1)
+	if distanceToOrigin > rangeThreshold then
+		return true
+	else
+		return false
+	end
+end
+
+function gadget:ProjectileCreated(proID, proOwnerID, weaponDefID)
+	if defWatchTable[weaponDefID] then
+		local originX, originy, originZ = spGetUnitPosition(proOwnerID)
+		lazyProjectileWatch[proID] = {weaponDefID = weaponDefID, proOwnerID = proOwnerID, originX = originX, originZ = originZ}
+	end
+end
+
+function gadget:GameFrame(frame)
+	if frame % edgyUpdateFrames == 2 then
+		for proID, proData in pairs(edgyProjectileWatch) do
+			local projectileX, projectileY, projectileZ = spGetProjectilePosition(proID)
+			if projectileX then
+				local defData = defWatchTable[proData.weaponDefID]
+				if projectileOverRangeCheck(proData.proOwnerID, defData.overRange, defData.leashRange, proData.originX, proData.originZ, projectileX, projectileZ) then
+					if defData.explodeMethod then
+						spSetProjectileCollision(proID)
+						edgyProjectileWatch[proID] = nil
+					elseif defData.descentMethod then
+						forcedDescentTable[proID] = descentSpeedStartingMultiplier
+						edgyProjectileWatch[proID] = nil
+					else
+						Spring.Echo("invalid destruction method")
+					end
+				end
+			else
+				edgyProjectileWatch[proID] = nil -- remove destroyed projectiles
+			end
+		end
+	end
+
+	if frame % lazyUpdateFrames == 3 then
+		for proID, proData in pairs(lazyProjectileWatch) do
+			local projectileX, projectileY, projectileZ = spGetProjectilePosition(proID)
+			if projectileX then
+				local defData = defWatchTable[proData.weaponDefID]
+				if projectileIsCloseToEdge(defData.rangeThreshold, proData.originX, proData.originZ, projectileX, projectileZ) then
+					edgyProjectileWatch[proID] = proData
+					lazyProjectileWatch[proID] = nil
+				end
+			else
+				lazyProjectileWatch[proID] = nil -- remove destroyed projectiles
+			end
+		end
+	end
+
+	if frame % forcedDescentUpdateFrames == 4 then
+		for proID, descentMultiplier in pairs(forcedDescentTable) do
+			local velocityX, velocityY, velocityZ, velocityW = spGetProjectileVelocity(proID)
+			if velocityY then
+				local newVelocityY = mathMax(velocityY - velocityW * descentMultiplier, -velocityW)
```

Comment:
this part should have at least a comment since it's doing a bit weird math

Reply:
rewrote the weird math for performance reasons, comment probably not necessary anymore